daq/host.py

```diff
@@ -300,20 +321,49 @@ def _main_module_timeout_handler(self):
         self.test_host = None
         self._docker_callback(exception=self._TIMEOUT_EXCEPTION)
 
-    def check_module_timeout(self):
-        """Checks module run time for each event loop"""
-        timeout_sec = self._get_test_timeout(self.test_name)
-        if not timeout_sec or not self.test_start:
+    def _check_module_timeout(self, test_name, test_start):
+        timeout_sec = self._get_test_timeout(test_name)
+        if not timeout_sec or not test_start:
             return
-        timeout = gcp.parse_timestamp(self.test_start) + timedelta(seconds=timeout_sec)
+        timeout = gcp.parse_timestamp(test_start) + timedelta(seconds=timeout_sec)
         nowtime = gcp.parse_timestamp(gcp.get_timestamp())
         if nowtime >= timeout:
             if self.timeout_handler:
-                LOGGER.error('Monitoring timeout for %s after %ds', self.test_name, timeout_sec)
+                LOGGER.error('Monitoring timeout for %s after %ds', test_name, timeout_sec)
                 # ensure it's called once
                 handler, self.timeout_handler = self.timeout_handler, None
                 handler()
 
+    def heartbeat(self):
+        """Checks module run time for each event loop"""
+        if self._active_deferrable_checks is not None:
+            self.timeout_handler = self._aux_module_timeout_handler
+            for check in self._active_deferrable_checks:
+                name, handler = check["name"], check["verify"]
+                self.test_name = name
+                # Several checks are happening each heartbeat so
+                # Timeout case needs to be handled separately
+                if "test_start" not in check:
+                    self.record_result(name, state=MODE.EXEC)
+                    check["test_start"] = gcp.get_timestamp()
+                self._check_module_timeout(self.test_name, check["test_start"])
+                try:
+                    result = handler(self)
+                    if result is not None:
+                        self.record_result(name, state=(MODE.DONE if result else MODE.MERR))
+                        check["done"] = True
+                except Exception as e:
+                    self.record_result(name, exception=e)
+                    self._report.accumulate(self.test_name, {ResultType.EXCEPTION: str(e)})
+            self._active_deferrable_checks = list(filter(lambda check: "done" not in check, \
```

Comment:
what is this line doing?  The problem with the list(fliter( stuff, although more concise, is it's often very obtuse...